### PR TITLE
Implement Normalize_Vector

### PR DESCRIPTION
**Pull request by 12345swordy**
Trying to implement #1628 for @sprunk

### CODE REVIEW

File: rts/Lua/LuaMathExtra.cpp [hunk 1]
```diff
@@ -184,6 +186,42 @@ int LuaMathExtra::smoothstep(lua_State* L) {
...
+
+/*** Returns the length of an given vector.
+ *
+ * @function math.vectorlength
+ * @number x
+ * @number y
+ * @number z
+ * @treturn vector length
+ */
+int LuaMathExtra::vectorlength(lua_State* L)
```
Comment (sprunk) [inline rts/Lua/LuaMathExtra.cpp:198]: seems redundant with `math.diag`.

File: rts/Lua/LuaMathExtra.cpp [hunk 2]
```diff
@@ -184,6 +186,42 @@ int LuaMathExtra::smoothstep(lua_State* L) {
...
+{
+	RECOIL_DETAILED_TRACY_ZONE;
+	float3 float3(luaL_checkfloat(L, 1), luaL_checkfloat(L, 2), luaL_checkfloat(L, 3));
+	lua_pushnumber(L, float3.Length());
+	return 1;
+}
+
+/*** Returns the normalize vector of an given vector.
+ *
+ * @function math.vectorvector
```
Comment (sprunk) [inline rts/Lua/LuaMathExtra.cpp:208]:
```suggestion
 * @function math.normalizevector
```

File: rts/Lua/LuaMathExtra.cpp [hunk 3]
```diff
@@ -184,6 +186,42 @@ int LuaMathExtra::smoothstep(lua_State* L) {
...
+ * @number x
+ * @number y
+ * @number z
+ * @treturn normalize vector
+ */
+int LuaMathExtra::normalizevector(lua_State* L)
```
Comment (sprunk) [inline rts/Lua/LuaMathExtra.cpp:214]: Should handle any dimensional vector, not just 3.

File: rts/Lua/LuaMathExtra.cpp (hunk 2 above)
Comment (12345swordy) [reply rts/Lua/LuaMathExtra.cpp]: Done

File: rts/Lua/LuaMathExtra.cpp (hunk 1 above)
Comment (12345swordy) [reply rts/Lua/LuaMathExtra.cpp]: You are right, I have deleted that.

File: rts/Lua/LuaMathExtra.cpp (hunk 3 above)
Comment (12345swordy) [reply rts/Lua/LuaMathExtra.cpp:197]: Done

File: rts/Lua/LuaMathExtra.cpp [hunk 4]
```diff
@@ -184,6 +185,34 @@ int LuaMathExtra::smoothstep(lua_State* L) {
...
+{
+	RECOIL_DETAILED_TRACY_ZONE;
+	lua_Number res = 0.0f;
+
+	for (int i = lua_gettop(L); i >= 1; i--) {
+		res += Square(luaL_checknumber_noassert(L, i));
+	}
+	res = math::sqrt(res);
+	if (res == 0)
```
Comment (sprunk) [inline rts/Lua/LuaMathExtra.cpp:206]: Should use the safe float check like this.
https://github.com/beyond-all-reason/spring/blob/c1e26101fb2e30ef57caa0fcfd496aec10d90af1/rts/System/Sync/SyncedFloat3.h#L469

File: rts/Lua/LuaMathExtra.cpp [hunk 5]
```diff
@@ -184,6 +185,34 @@ int LuaMathExtra::smoothstep(lua_State* L) {
...
+		return 0;
```
Comment (sprunk) [inline rts/Lua/LuaMathExtra.cpp:207]: Should push a zero-vector and not nil.

File: rts/Lua/LuaMathExtra.cpp [hunk 6]
```diff
@@ -184,6 +185,34 @@ int LuaMathExtra::smoothstep(lua_State* L) {
...
+	for (int i = lua_gettop(L); i >= 1; i--) {
+		float tmp = luaL_checknumber_noassert(L, i);
+		tmp = res / tmp;
```
Comment (sprunk) [inline rts/Lua/LuaMathExtra.cpp:210]: Looks incorrect.

File: rts/Lua/LuaMathExtra.cpp [hunk 7]
```diff
@@ -184,6 +185,34 @@ int LuaMathExtra::smoothstep(lua_State* L) {
...
+		lua_pushnumber(L, tmp);
+	}
+	return lua_gettop(L);
```
Comment (sprunk) [inline rts/Lua/LuaMathExtra.cpp:213]: Looks incorrect.

File: rts/Lua/LuaMathExtra.cpp [hunk 8]
```diff
@@ -184,6 +185,34 @@ int LuaMathExtra::smoothstep(lua_State* L) {
...
 }
 
+
+/*** Returns the normalize vector of an given vector.
+ *
+ * @function math.normalize
+ * @number x
+ * @number y
+ * @number z
+ * @treturn nil|normalize vector
```
Comment (sprunk) [inline rts/Lua/LuaMathExtra.cpp:195]: Accepts and returns N numbers. Check out the docs for `math.diag` for an example.

**Review by sprunk — CHANGES_REQUESTED** [review]

File: rts/Lua/LuaMathExtra.cpp (hunk 5 above)
Comment (12345swordy) [reply rts/Lua/LuaMathExtra.cpp:207]: That does not make any mathematical sense as you can't divide by zero.

File: rts/Lua/LuaMathExtra.cpp (hunk 6 above)
Comment (12345swordy) [reply rts/Lua/LuaMathExtra.cpp:210]: You're right it should be tmp = tmp / res. It fix now

File: rts/Lua/LuaMathExtra.cpp (hunk 5 above)
Comment (12345swordy) [reply rts/Lua/LuaMathExtra.cpp:207]: I will do it, but it doesn't change the fact that it a mathematical oxymoronic.

File: rts/Lua/LuaMathExtra.cpp [hunk 9]
```diff
@@ -26,6 +26,7 @@ bool LuaMathExtra::PushEntries(lua_State* L)
 	LuaPushNamedCFunc(L, "round",  round);
 	LuaPushNamedCFunc(L, "erf",    erf);
 	LuaPushNamedCFunc(L, "smoothstep", smoothstep);
+	LuaPushNamedCFunc(L, "normalizevector", normalizevector);
```
Comment (sprunk) [inline rts/Lua/LuaMathExtra.cpp:29]:
```suggestion
	LuaPushNamedCFunc(L, "normalize", normalizevector);
```

File: rts/Lua/LuaMathExtra.cpp [hunk 10]
```diff
@@ -184,6 +185,39 @@ int LuaMathExtra::smoothstep(lua_State* L) {
 	return 1;
 }
 
+
+/*** Returns the normalize vector of an given vector.
+ *
+ * @function math.normalize
+ * @number x
+ * @number[opt] xn and so on
+ * @treturn zero xn numbers|xn numbers
```
Comment (sprunk) [inline rts/Lua/LuaMathExtra.cpp:194]:
```suggestion
 * @number x1
 * @number[opt] x2
 * @number[opt] xN and so on
 * @treturn number normalized1
 * @treturn number normalized2 and so on
```

File: rts/Lua/LuaMathExtra.cpp [hunk 11]
```diff
@@ -184,6 +185,39 @@ int LuaMathExtra::smoothstep(lua_State* L) {
...
+{
+	RECOIL_DETAILED_TRACY_ZONE;
+	lua_Number res = 0.0f;
+	const int param = lua_gettop(L);
+	for (int i = param; i >= 1; i--) {
+		res += Square(luaL_checknumber_noassert(L, i));
+	}
+	if likely(res > float3::nrm_eps())
+		res = math::sqrt(res);
+	if (res == 0)
```
Comment (sprunk) [inline rts/Lua/LuaMathExtra.cpp:206]:
```suggestion
	else
```

File: rts/Lua/LuaMathExtra.cpp [hunk 12]
```diff
@@ -184,6 +185,39 @@ int LuaMathExtra::smoothstep(lua_State* L) {
...
+	{
+		for (int i = param; i >= 1; i--) {
+			lua_pushnumber(L, 0);
+		}
+		return param;
+	}
+	for (int i = param; i >= 1; i--) {
```
Comment (sprunk) [inline rts/Lua/LuaMathExtra.cpp:213]: Currently pushes backwards
```suggestion
	for (int i = 1; <= param; ++i) {
```

**Review by sprunk — CHANGES_REQUESTED** [review]
Almost works!

**Review by sprunk — APPROVED** [review]

**Pull request merged**